{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import igraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cdr3info = pd.read_table(\"input_files/nlv-llw.txt\")\n",
    "cdr3graph = pd.read_table(\"input_files/nlv-llw.graph.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdr3aa</th>\n",
       "      <th>epitope</th>\n",
       "      <th>species</th>\n",
       "      <th>gene</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>v.end</th>\n",
       "      <th>j.start</th>\n",
       "      <th>cid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAAADDKIIF</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>h.A.LLWNGPMAV.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAAFDDKIIF</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>h.A.LLWNGPMAV.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAAIDDKIIF</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>h.A.LLWNGPMAV.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAAKDDKIIF</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>h.A.LLWNGPMAV.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAALDDKIIF</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>h.A.LLWNGPMAV.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cdr3aa    epitope      species gene       v.segm     j.segm  v.end  \\\n",
       "0  CAAADDKIIF  LLWNGPMAV  HomoSapiens  TRA  TRAV12-2*01  TRAJ30*01      2   \n",
       "1  CAAFDDKIIF  LLWNGPMAV  HomoSapiens  TRA  TRAV12-2*01  TRAJ30*01      2   \n",
       "2  CAAIDDKIIF  LLWNGPMAV  HomoSapiens  TRA  TRAV12-2*01  TRAJ30*01      2   \n",
       "3  CAAKDDKIIF  LLWNGPMAV  HomoSapiens  TRA  TRAV12-2*01  TRAJ30*01      2   \n",
       "4  CAALDDKIIF  LLWNGPMAV  HomoSapiens  TRA  TRAV12-2*01  TRAJ30*01      2   \n",
       "\n",
       "   j.start              cid  \n",
       "0        4  h.A.LLWNGPMAV.1  \n",
       "1        4  h.A.LLWNGPMAV.1  \n",
       "2        4  h.A.LLWNGPMAV.1  \n",
       "3        4  h.A.LLWNGPMAV.1  \n",
       "4        4  h.A.LLWNGPMAV.1  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdr3info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene</th>\n",
       "      <th>epitope</th>\n",
       "      <th>to.cdr3aa</th>\n",
       "      <th>from.cdr3aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>human</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVSSDKIIF</td>\n",
       "      <td>CAVSGDKIIF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>human</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVSTDKLIF</td>\n",
       "      <td>CAVNTDKLIF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVTDDKIIF</td>\n",
       "      <td>CAVADDKIIF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>human</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVTDDKIIF</td>\n",
       "      <td>CAATDDKIIF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVSTDKLIF</td>\n",
       "      <td>CAVSPDKLIF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species gene    epitope   to.cdr3aa from.cdr3aa\n",
       "0   human  TRA  LLWNGPMAV  CAVSSDKIIF  CAVSGDKIIF\n",
       "1   human  TRA  LLWNGPMAV  CAVSTDKLIF  CAVNTDKLIF\n",
       "2   human  TRA  LLWNGPMAV  CAVTDDKIIF  CAVADDKIIF\n",
       "3   human  TRA  LLWNGPMAV  CAVTDDKIIF  CAATDDKIIF\n",
       "4   human  TRA  LLWNGPMAV  CAVSTDKLIF  CAVSPDKLIF"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdr3graph.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "147\n",
      "147\n"
     ]
    }
   ],
   "source": [
    "#assertion, that there is no cdr3 specific to two epitopes.\n",
    "print(len(cdr3info))\n",
    "print(len(cdr3info.drop_duplicates(subset=[\"cdr3aa\", \"v.segm\"], keep=\"first\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['cdr3aa', 'epitope', 'species', 'gene', 'v.segm', 'j.segm', 'v.end',\n",
      "       'j.start', 'cid'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(cdr3info.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cdr3info.rename(columns={\"cdr3aa\":\"from.cdr3aa\", \n",
    "                         \"v.segm\":\"from.v.segm\", \n",
    "                         \"j.segm\":\"from.j.segm\",\n",
    "                         \"v.end\":\"from.v.end\",\n",
    "                         \"j.start\":\"from.j.start\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1600"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cdr3graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene</th>\n",
       "      <th>epitope</th>\n",
       "      <th>to.cdr3aa</th>\n",
       "      <th>from.cdr3aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVSSDKIIF</td>\n",
       "      <td>CAVSGDKIIF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVSTDKLIF</td>\n",
       "      <td>CAVNTDKLIF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVTDDKIIF</td>\n",
       "      <td>CAVADDKIIF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVTDDKIIF</td>\n",
       "      <td>CAATDDKIIF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVSTDKLIF</td>\n",
       "      <td>CAVSPDKLIF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       species gene    epitope   to.cdr3aa from.cdr3aa\n",
       "0  HomoSapiens  TRA  LLWNGPMAV  CAVSSDKIIF  CAVSGDKIIF\n",
       "1  HomoSapiens  TRA  LLWNGPMAV  CAVSTDKLIF  CAVNTDKLIF\n",
       "2  HomoSapiens  TRA  LLWNGPMAV  CAVTDDKIIF  CAVADDKIIF\n",
       "3  HomoSapiens  TRA  LLWNGPMAV  CAVTDDKIIF  CAATDDKIIF\n",
       "4  HomoSapiens  TRA  LLWNGPMAV  CAVSTDKLIF  CAVSPDKLIF"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spec = {\"human\":\"HomoSapiens\"}\n",
    "cdr3graph.replace({\"species\":spec}, inplace=True)\n",
    "cdr3graph.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cdr3graph = cdr3graph.merge(cdr3info, on=[\"species\", \"gene\", \"epitope\", \"from.cdr3aa\"], how=\"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1600"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cdr3graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cdr3info.rename(columns={\"from.cdr3aa\":\"to.cdr3aa\", \n",
    "                         \"from.v.segm\":\"to.v.segm\", \n",
    "                         \"from.j.segm\":\"to.j.segm\",\n",
    "                         \"from.v.end\":\"to.v.end\",\n",
    "                         \"from.j.start\":\"to.j.start\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "cdr3graph = cdr3graph.merge(cdr3info, on=[\"species\", \"gene\", \"epitope\", \"to.cdr3aa\", \"cid\"], how=\"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1600"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cdr3graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       species gene    epitope   to.cdr3aa  from.cdr3aa  from.v.segm  \\\n",
      "0  HomoSapiens  TRA  LLWNGPMAV  CAVSSDKIIF   CAVSGDKIIF  TRAV12-2*01   \n",
      "1  HomoSapiens  TRA  LLWNGPMAV  CAVSSDKIIF   CAVSDDKIIF  TRAV12-2*01   \n",
      "2  HomoSapiens  TRA  LLWNGPMAV  CAVSSDKIIF   CAVGSDKIIF  TRAV12-2*01   \n",
      "3  HomoSapiens  TRA  LLWNGPMAV  CAVTGDKIIF   CAVSGDKIIF  TRAV12-2*01   \n",
      "4  HomoSapiens  TRA  LLWNGPMAV  CAVTGDKIIF  CAVTGDDKIIF  TRAV12-2*01   \n",
      "\n",
      "  from.j.segm  from.v.end  from.j.start              cid    to.v.segm  \\\n",
      "0   TRAJ30*01         3.0           5.0  h.A.LLWNGPMAV.1  TRAV12-2*01   \n",
      "1   TRAJ30*01         3.0           4.0  h.A.LLWNGPMAV.1  TRAV12-2*01   \n",
      "2   TRAJ30*01         3.0           5.0  h.A.LLWNGPMAV.1  TRAV12-2*01   \n",
      "3   TRAJ30*01         3.0           5.0  h.A.LLWNGPMAV.1  TRAV12-2*01   \n",
      "4   TRAJ30*01         3.0           5.0  h.A.LLWNGPMAV.1  TRAV12-2*01   \n",
      "\n",
      "   to.j.segm  to.v.end  to.j.start  \n",
      "0  TRAJ30*01       3.0         5.0  \n",
      "1  TRAJ30*01       3.0         5.0  \n",
      "2  TRAJ30*01       3.0         5.0  \n",
      "3  TRAJ30*01       3.0         5.0  \n",
      "4  TRAJ30*01       3.0         5.0  \n"
     ]
    }
   ],
   "source": [
    "print(cdr3graph.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def tuple_maker(df, tuplecol, *values):\n",
    "    def gettuple(*values):\n",
    "        return sorted(*values)\n",
    "    \n",
    "    df[tuplecol] = df.apply(lambda col: gettuple(col[val] for val in values), axis=1)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cdr3graph = tuple_maker(cdr3graph, \"tuple\", \"to.cdr3aa\", \"from.cdr3aa\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cdr3graph_missing = cdr3graph[cdr3graph[\"v.segm\"].isna()]\n",
    "cdr3graph_missing.to_csv(\"input_files/nlv-llw.missing_values.txt\", \n",
    "sep=\"\\t\", \n",
    "index=None)\n",
    "\n",
    "cdr3info[\"cid\"].unique()\n",
    "\n",
    "cdr3graph = tuple_maker(cdr3graph, \n",
    "\"tuple\", \n",
    "\"to.cdr3aa\", \n",
    "\"from.cdr3aa\")\n",
    "\n",
    "graph = igraph.Graph.TupleList(list(cdr3graph[\"tuple\"]))\n",
    "\n",
    "vertices = graph.vs['name']\n",
    "\n",
    "vertices_unique = list(cdr3info.drop_duplicates(subset=[\"cid\"], \n",
    "keep=\"first\")[\"to.cdr3aa\"])\n",
    "\n",
    "vertices_unique_positions = np.where(np.isin(vertices, vertices_unique))[0]\n",
    "\n",
    "print(list((vertices[i] in vertices_unique) for i in vertices_unique_positions))\n",
    "\n",
    "edge_distances = graph.shortest_paths_dijkstra()\n",
    "\n",
    "for i in vertices_unique_positions:\n",
    "    edges = np.array(edge_distances[i])\n",
    "    print(len(np.where(~np.isin(edges, np.inf))[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cdr3graph_work = cdr3graph[~(cdr3graph[\"cid\"].isna())].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "556 278 556\n"
     ]
    }
   ],
   "source": [
    "print(len(cdr3graph_work), #check num of edges\n",
    "      \n",
    "      #check num of unique edges in undirected graph;\n",
    "      len(cdr3graph_work[\"tuple\"].str.join(\"|\").drop_duplicates(keep=\"first\")),\n",
    "      \n",
    "      #check num of unique edges #in directed graph)\n",
    "      len(cdr3graph_work.drop_duplicates(subset=[\"to.cdr3aa\",\n",
    "                                                 \"from.cdr3aa\"],\n",
    "                                         keep=\"first\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def bfs_from_minmean(cluster_graph):\n",
    "    work_graph = pd.DataFrame()\n",
    "    vertices = cluster_graph.vs['name']\n",
    "    edge_distances = cluster_graph.shortest_paths_dijkstra()#mode=\"ALL\")\n",
    "    mean_distances = [np.mean(i) for i in edge_distances]\n",
    "    \n",
    "    vertix_with_minmean = np.argmin(mean_distances)\n",
    "    \n",
    "    checked_vertices = [vertix_with_minmean]\n",
    "    \n",
    "    while len(checked_vertices) != len(vertices):\n",
    "        for vertix in checked_vertices:\n",
    "            edges = edge_distances[vertix]\n",
    "            nb_vertices = list(np.argwhere(np.array(edges) == 1)[:, 0])\n",
    "            for new_vertix in nb_vertices:\n",
    "                if new_vertix in checked_vertices:\n",
    "                    continue\n",
    "\n",
    "                from_vertix = vertices[vertix]\n",
    "                to_vertix = vertices[new_vertix]\n",
    "                work_graph = work_graph.append(cdrs[(cdrs[\"from.cdr3aa\"] == from_vertix) & (\n",
    "                        cdrs[\"to.cdr3aa\"] == to_vertix)])\n",
    "                checked_vertices.append(new_vertix)\n",
    "            \n",
    "    return work_graph\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sorted_graph = pd.DataFrame()\n",
    "for cluster, cdrs in cdr3graph_work.groupby(\"cid\"):\n",
    "    cluster_graph = igraph.Graph.TupleList(list(cdrs[\"tuple\"]))\n",
    "    sorted_graph = sorted_graph.append(bfs_from_minmean(cluster_graph))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sorted_graph = sorted_graph.reset_index(drop=True)\n",
    "sorted_graph.to_csv(\"nlv-llw.graphpath.txt\", sep=\"\\t\", index=None, columns=sorted_graph.columns[:-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we have sorted graph.\n",
    "\n",
    "Time for central vertices: we just need to get first vertices for each cluster in sorted_graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene</th>\n",
       "      <th>epitope</th>\n",
       "      <th>to.cdr3aa</th>\n",
       "      <th>from.cdr3aa</th>\n",
       "      <th>from.v.segm</th>\n",
       "      <th>from.j.segm</th>\n",
       "      <th>from.v.end</th>\n",
       "      <th>from.j.start</th>\n",
       "      <th>cid</th>\n",
       "      <th>to.v.segm</th>\n",
       "      <th>to.j.segm</th>\n",
       "      <th>to.v.end</th>\n",
       "      <th>to.j.start</th>\n",
       "      <th>tuple</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAVSDDKIIF</td>\n",
       "      <td>CAVTDDKIIF</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>h.A.LLWNGPMAV.1</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[CAVSDDKIIF, CAVTDDKIIF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CAGSNARLMF</td>\n",
       "      <td>CAVSNARLMF</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>h.A.LLWNGPMAV.13</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[CAGSNARLMF, CAVSNARLMF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>CASPTGNQFYF</td>\n",
       "      <td>CASNTGNQFYF</td>\n",
       "      <td>TRAV24*01,TRAV18*01</td>\n",
       "      <td>TRAJ49*01</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>h.A.NLVPMVATV.15</td>\n",
       "      <td>TRAV24*01</td>\n",
       "      <td>TRAJ49*01</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[CASNTGNQFYF, CASPTGNQFYF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>CFLDNNNDMRF</td>\n",
       "      <td>CILDNNNDMRF</td>\n",
       "      <td>TRAV26-2*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>h.A.NLVPMVATV.51</td>\n",
       "      <td>TRAV26-2*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[CFLDNNNDMRF, CILDNNNDMRF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRB</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CATSGGGAYEQYF</td>\n",
       "      <td>CATSRGGAYEQYF</td>\n",
       "      <td>TRBV15*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>h.B.LLWNGPMAV.14</td>\n",
       "      <td>TRBV15*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>[CATSGGGAYEQYF, CATSRGGAYEQYF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRB</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>CASSAGTGAYEQYF</td>\n",
       "      <td>CASSPGTGAYEQYF</td>\n",
       "      <td>TRBV27*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>h.B.LLWNGPMAV.3</td>\n",
       "      <td>TRBV9*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>[CASSAGTGAYEQYF, CASSPGTGAYEQYF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRB</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>CASSYQTGASYGYTF</td>\n",
       "      <td>CASSYQTGAIYGYTF</td>\n",
       "      <td>TRBV6-5*01</td>\n",
       "      <td>TRBJ1-2*01</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>h.B.NLVPMVATV.32</td>\n",
       "      <td>TRBV6-3*01,TRBV6-1*01,TRBV6-5*01</td>\n",
       "      <td>TRBJ1-2*01</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>[CASSYQTGAIYGYTF, CASSYQTGASYGYTF]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRB</td>\n",
       "      <td>NLVPMVATV</td>\n",
       "      <td>CASSLPPGATNEKLFF</td>\n",
       "      <td>CASSLAPGATNEKLFF</td>\n",
       "      <td>TRBV7-9*01,TRBV6-1*01,TRBV7-8*01,TRBV7-6*01,TR...</td>\n",
       "      <td>TRBJ1-4*01</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>h.B.NLVPMVATV.6</td>\n",
       "      <td>TRBV7-6*01</td>\n",
       "      <td>TRBJ1-4*01</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>[CASSLAPGATNEKLFF, CASSLPPGATNEKLFF]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         species gene    epitope         to.cdr3aa       from.cdr3aa  \\\n",
       "0    HomoSapiens  TRA  LLWNGPMAV        CAVSDDKIIF        CAVTDDKIIF   \n",
       "49   HomoSapiens  TRA  LLWNGPMAV        CAGSNARLMF        CAVSNARLMF   \n",
       "55   HomoSapiens  TRA  NLVPMVATV       CASPTGNQFYF       CASNTGNQFYF   \n",
       "65   HomoSapiens  TRA  NLVPMVATV       CFLDNNNDMRF       CILDNNNDMRF   \n",
       "82   HomoSapiens  TRB  LLWNGPMAV     CATSGGGAYEQYF     CATSRGGAYEQYF   \n",
       "86   HomoSapiens  TRB  LLWNGPMAV    CASSAGTGAYEQYF    CASSPGTGAYEQYF   \n",
       "104  HomoSapiens  TRB  NLVPMVATV   CASSYQTGASYGYTF   CASSYQTGAIYGYTF   \n",
       "116  HomoSapiens  TRB  NLVPMVATV  CASSLPPGATNEKLFF  CASSLAPGATNEKLFF   \n",
       "\n",
       "                                           from.v.segm from.j.segm  \\\n",
       "0                                          TRAV12-2*01   TRAJ30*01   \n",
       "49                                         TRAV12-2*01   TRAJ31*01   \n",
       "55                                 TRAV24*01,TRAV18*01   TRAJ49*01   \n",
       "65                                         TRAV26-2*01   TRAJ43*01   \n",
       "82                                           TRBV15*01  TRBJ2-7*01   \n",
       "86                                           TRBV27*01  TRBJ2-7*01   \n",
       "104                                         TRBV6-5*01  TRBJ1-2*01   \n",
       "116  TRBV7-9*01,TRBV6-1*01,TRBV7-8*01,TRBV7-6*01,TR...  TRBJ1-4*01   \n",
       "\n",
       "     from.v.end  from.j.start               cid  \\\n",
       "0           3.0           4.0   h.A.LLWNGPMAV.1   \n",
       "49          3.0           4.0  h.A.LLWNGPMAV.13   \n",
       "55          2.0           3.0  h.A.NLVPMVATV.15   \n",
       "65          3.0           4.0  h.A.NLVPMVATV.51   \n",
       "82          5.0           8.0  h.B.LLWNGPMAV.14   \n",
       "86          4.0           9.0   h.B.LLWNGPMAV.3   \n",
       "104         5.0          10.0  h.B.NLVPMVATV.32   \n",
       "116         5.0           9.0   h.B.NLVPMVATV.6   \n",
       "\n",
       "                            to.v.segm   to.j.segm  to.v.end  to.j.start  \\\n",
       "0                         TRAV12-2*01   TRAJ30*01       3.0         4.0   \n",
       "49                        TRAV12-2*01   TRAJ31*01       2.0         4.0   \n",
       "55                          TRAV24*01   TRAJ49*01       2.0         4.0   \n",
       "65                        TRAV26-2*01   TRAJ43*01      -1.0         4.0   \n",
       "82                          TRBV15*01  TRBJ2-7*01       4.0         8.0   \n",
       "86                           TRBV9*01  TRBJ2-7*01       4.0         9.0   \n",
       "104  TRBV6-3*01,TRBV6-1*01,TRBV6-5*01  TRBJ1-2*01       5.0        10.0   \n",
       "116                        TRBV7-6*01  TRBJ1-4*01       5.0         9.0   \n",
       "\n",
       "                                    tuple  \n",
       "0                [CAVSDDKIIF, CAVTDDKIIF]  \n",
       "49               [CAGSNARLMF, CAVSNARLMF]  \n",
       "55             [CASNTGNQFYF, CASPTGNQFYF]  \n",
       "65             [CFLDNNNDMRF, CILDNNNDMRF]  \n",
       "82         [CATSGGGAYEQYF, CATSRGGAYEQYF]  \n",
       "86       [CASSAGTGAYEQYF, CASSPGTGAYEQYF]  \n",
       "104    [CASSYQTGAIYGYTF, CASSYQTGASYGYTF]  \n",
       "116  [CASSLAPGATNEKLFF, CASSLPPGATNEKLFF]  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "minmeanvertices = sorted_graph.drop_duplicates(subset=\"cid\", keep=\"first\")\n",
    "minmeanvertices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene</th>\n",
       "      <th>epitope</th>\n",
       "      <th>to.cdr3aa</th>\n",
       "      <th>from.cdr3aa</th>\n",
       "      <th>from.v.segm</th>\n",
       "      <th>from.j.segm</th>\n",
       "      <th>from.v.end</th>\n",
       "      <th>from.j.start</th>\n",
       "      <th>cid</th>\n",
       "      <th>to.v.segm</th>\n",
       "      <th>to.j.segm</th>\n",
       "      <th>to.v.end</th>\n",
       "      <th>to.j.start</th>\n",
       "      <th>tuple</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [species, gene, epitope, to.cdr3aa, from.cdr3aa, from.v.segm, from.j.segm, from.v.end, from.j.start, cid, to.v.segm, to.j.segm, to.v.end, to.j.start, tuple]\n",
       "Index: []"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "minmeanvertices[minmeanvertices[\"from.v.end\"]<0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Uh oh\n",
    "#Strange! Let's take a look in segments.aaparts.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>type</th>\n",
       "      <th>segm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>TRAV26-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CA</td>\n",
       "      <td>V</td>\n",
       "      <td>TRAV24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAA</td>\n",
       "      <td>V</td>\n",
       "      <td>TRAV13-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAAS</td>\n",
       "      <td>V</td>\n",
       "      <td>TRAV29/DV5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAE</td>\n",
       "      <td>V</td>\n",
       "      <td>TRAV5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       species gene  cdr3 type        segm\n",
       "0  HomoSapiens  TRA     C    V    TRAV26-1\n",
       "1  HomoSapiens  TRA    CA    V      TRAV24\n",
       "2  HomoSapiens  TRA   CAA    V    TRAV13-1\n",
       "3  HomoSapiens  TRA  CAAS    V  TRAV29/DV5\n",
       "4  HomoSapiens  TRA   CAE    V       TRAV5"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#segment_parts = pd.read_table(\"segments.aaparts.txt\")\n",
    "#segment_parts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#strange_tcr = minmeanvertices[minmeanvertices[\"from.v.end\"]<0]\n",
    "#strange_cdr3 = list(strange_tcr[\"from.cdr3aa\"])[0]\n",
    "#strange_vsegm = list(strange_tcr[\"from.v.segm\"])[0]\n",
    "#strange_specie = list(strange_tcr[\"species\"])[0]\n",
    "#print(strange_cdr3, strange_vsegm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sliced_cdr3 = strange_cdr3\n",
    "#while len(sliced_cdr3) != 0:\n",
    "#    if len(segment_parts[(segment_parts[\"species\"] == strange_specie) &\n",
    "#                         (segment_parts[\"cdr3\"] == sliced_cdr3) & \n",
    "#                         (segment_parts[\"segm\"]+\"*01\" == strange_vsegm)]):\n",
    "#        break\n",
    "#    sliced_cdr3 = sliced_cdr3[:-1]\n",
    "#print(sliced_cdr3, len(sliced_cdr3))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Anyway, let's create aminoacid sequences of selected tcrs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gene_to_protein\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcr_n_constructor = gene_to_protein.tcr_nucleotide_constructor()\n",
    "tcr_a_constructor = gene_to_protein.tcr_aminoacid_constructor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.0 4.0 CAVTDDKIIF ['TRAV12-2*01'] ['TRAJ30*01']\n",
      "QKEVEQNSGPLSVPEGAIASLNCTYSDRGSQSFFWYRQYSGKSPELIMFIYSNGDKEDGRFTAQLNKASQYVSLLIRDSQPSDSATYL\n",
      "QKEVEQNSGPLSVPEGAIASLNCTYSDRGSQSFFWYRQYSGKSPELIMFIYSNGDKEDGRFTAQLNKASQYVSLLIRDSQPSDSATYLCAVXDDKIIF\n",
      "QKEVEQNSGPLSVPEGAIASLNCTYSDRGSQSFFWYRQYSGKSPELIMFIYSNGDKEDGRFTAQLNKASQYVSLLIRDSQPSDSATYLCAVXDDKIIFGKGTRLHILP\n",
      "QKEVEQNSGPLSVPEGAIASLNCTYSDRGSQSFFWYRQYSGKSPELIMFIYSNGDKEDGRFTAQLNKASQYVSLLIRDSQPSDSATYLCAVXDDKIIFGKGTRLHILPNIQNPDPAVYQLRDSKSSDKSVCLFTDFDSQTNVSQSKDSDVYITDKTVLDMRSMDFKSNSAVAWSNKSDFACANAFNNSIIPEDTFFPSPESSCDVKLVEKSFETDTNLNFQNLSVIGFRILLLKVAGFNLLMTLRLWSS*\n",
      "3.0 4.0 CAVSNARLMF ['TRAV12-2*01'] ['TRAJ31*01']\n",
      "QKEVEQNSGPLSVPEGAIASLNCTYSDRGSQSFFWYRQYSGKSPELIMFIYSNGDKEDGRFTAQLNKASQYVSLLIRDSQPSDSATYL\n",
      "QKEVEQNSGPLSVPEGAIASLNCTYSDRGSQSFFWYRQYSGKSPELIMFIYSNGDKEDGRFTAQLNKASQYVSLLIRDSQPSDSATYLCAVXNARLMF\n",
      "QKEVEQNSGPLSVPEGAIASLNCTYSDRGSQSFFWYRQYSGKSPELIMFIYSNGDKEDGRFTAQLNKASQYVSLLIRDSQPSDSATYLCAVXNARLMFGDGTQLVVKP\n",
      "QKEVEQNSGPLSVPEGAIASLNCTYSDRGSQSFFWYRQYSGKSPELIMFIYSNGDKEDGRFTAQLNKASQYVSLLIRDSQPSDSATYLCAVXNARLMFGDGTQLVVKPNIQNPDPAVYQLRDSKSSDKSVCLFTDFDSQTNVSQSKDSDVYITDKTVLDMRSMDFKSNSAVAWSNKSDFACANAFNNSIIPEDTFFPSPESSCDVKLVEKSFETDTNLNFQNLSVIGFRILLLKVAGFNLLMTLRLWSS*\n",
      "2.0 3.0 CASNTGNQFYF ['TRAV24*01', 'TRAV18*01'] ['TRAJ49*01']\n",
      "ILNVEQSPQSLHVQEGDSTNFTCSFPSSNFYALHWYRWETAKSPEALFVMTLNGDEKKKGRISATLNTKEGYSYLYIKGSQPEDSATYL\n",
      "ILNVEQSPQSLHVQEGDSTNFTCSFPSSNFYALHWYRWETAKSPEALFVMTLNGDEKKKGRISATLNTKEGYSYLYIKGSQPEDSATYLCAXNTGNQFYF\n",
      "ILNVEQSPQSLHVQEGDSTNFTCSFPSSNFYALHWYRWETAKSPEALFVMTLNGDEKKKGRISATLNTKEGYSYLYIKGSQPEDSATYLCAXNTGNQFYFGTGTSLTVIP\n",
      "ILNVEQSPQSLHVQEGDSTNFTCSFPSSNFYALHWYRWETAKSPEALFVMTLNGDEKKKGRISATLNTKEGYSYLYIKGSQPEDSATYLCAXNTGNQFYFGTGTSLTVIPNIQNPDPAVYQLRDSKSSDKSVCLFTDFDSQTNVSQSKDSDVYITDKTVLDMRSMDFKSNSAVAWSNKSDFACANAFNNSIIPEDTFFPSPESSCDVKLVEKSFETDTNLNFQNLSVIGFRILLLKVAGFNLLMTLRLWSS*\n",
      "2.0 3.0 CASNTGNQFYF ['TRAV24*01', 'TRAV18*01'] ['TRAJ49*01']\n",
      "GDSVTQTEGPVTLPERAALTLNCTYQSSYSTFLFWYVQYLNKEPELLLKSSENQETDSRGFQASPIKSDSSFHLEKPSVQLSDSAVYY\n",
      "GDSVTQTEGPVTLPERAALTLNCTYQSSYSTFLFWYVQYLNKEPELLLKSSENQETDSRGFQASPIKSDSSFHLEKPSVQLSDSAVYYCAXNTGNQFYF\n",
      "GDSVTQTEGPVTLPERAALTLNCTYQSSYSTFLFWYVQYLNKEPELLLKSSENQETDSRGFQASPIKSDSSFHLEKPSVQLSDSAVYYCAXNTGNQFYFGTGTSLTVIP\n",
      "GDSVTQTEGPVTLPERAALTLNCTYQSSYSTFLFWYVQYLNKEPELLLKSSENQETDSRGFQASPIKSDSSFHLEKPSVQLSDSAVYYCAXNTGNQFYFGTGTSLTVIPNIQNPDPAVYQLRDSKSSDKSVCLFTDFDSQTNVSQSKDSDVYITDKTVLDMRSMDFKSNSAVAWSNKSDFACANAFNNSIIPEDTFFPSPESSCDVKLVEKSFETDTNLNFQNLSVIGFRILLLKVAGFNLLMTLRLWSS*\n",
      "3.0 4.0 CILDNNNDMRF ['TRAV26-2*01'] ['TRAJ43*01']\n",
      "DAKTTQPNSMESNEEEPVHLPCNHSTISGTDYIHWYRQLPSQGPEYVIHGLTSNVNNRMASLAIAEDRKSSTLILHRATLRDAAVYY\n",
      "DAKTTQPNSMESNEEEPVHLPCNHSTISGTDYIHWYRQLPSQGPEYVIHGLTSNVNNRMASLAIAEDRKSSTLILHRATLRDAAVYYCILXNNNDMRF\n",
      "DAKTTQPNSMESNEEEPVHLPCNHSTISGTDYIHWYRQLPSQGPEYVIHGLTSNVNNRMASLAIAEDRKSSTLILHRATLRDAAVYYCILXNNNDMRFGAGTRLTVKP\n",
      "DAKTTQPNSMESNEEEPVHLPCNHSTISGTDYIHWYRQLPSQGPEYVIHGLTSNVNNRMASLAIAEDRKSSTLILHRATLRDAAVYYCILXNNNDMRFGAGTRLTVKPNIQNPDPAVYQLRDSKSSDKSVCLFTDFDSQTNVSQSKDSDVYITDKTVLDMRSMDFKSNSAVAWSNKSDFACANAFNNSIIPEDTFFPSPESSCDVKLVEKSFETDTNLNFQNLSVIGFRILLLKVAGFNLLMTLRLWSS*\n",
      "5.0 8.0 CATSRGGAYEQYF ['TRBV15*01'] ['TRBJ2-7*01']\n",
      "DAMVIQNPRYQVTQFGKPVTLSCSQTLNHNVMYWYQQKSSQAPKLLFHYYDKDFNNEADTPDNFQSRRPNTSFCFLDIRSPGLGDTAMYL\n",
      "DAMVIQNPRYQVTQFGKPVTLSCSQTLNHNVMYWYQQKSSQAPKLLFHYYDKDFNNEADTPDNFQSRRPNTSFCFLDIRSPGLGDTAMYLCATSRXXXYEQYF\n",
      "DAMVIQNPRYQVTQFGKPVTLSCSQTLNHNVMYWYQQKSSQAPKLLFHYYDKDFNNEADTPDNFQSRRPNTSFCFLDIRSPGLGDTAMYLCATSRXXXYEQYFGPGTRLTVT\n",
      "DAMVIQNPRYQVTQFGKPVTLSCSQTLNHNVMYWYQQKSSQAPKLLFHYYDKDFNNEADTPDNFQSRRPNTSFCFLDIRSPGLGDTAMYLCATSRXXXYEQYFGPGTRLTVTEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWWVNGKEVHSGVSTDPQPLKEQPALNDSRYCLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSESYQQGVLSATILYEILLGKATLYAVLVSALVLMAMVKRKDSRG\n",
      "4.0 9.0 CASSPGTGAYEQYF ['TRBV27*01'] ['TRBJ2-7*01']\n",
      "EAQVTQNPRYLITVTGKKLTVTCSQNMNHEYMSWYRQDPGLGLRQIYYSMNVEVTDKGDVPEGYKVSRKEKRNFPLILESPSPNQTSLYF\n",
      "EAQVTQNPRYLITVTGKKLTVTCSQNMNHEYMSWYRQDPGLGLRQIYYSMNVEVTDKGDVPEGYKVSRKEKRNFPLILESPSPNQTSLYFCASSXXXXXYEQYF\n",
      "EAQVTQNPRYLITVTGKKLTVTCSQNMNHEYMSWYRQDPGLGLRQIYYSMNVEVTDKGDVPEGYKVSRKEKRNFPLILESPSPNQTSLYFCASSXXXXXYEQYFGPGTRLTVT\n",
      "EAQVTQNPRYLITVTGKKLTVTCSQNMNHEYMSWYRQDPGLGLRQIYYSMNVEVTDKGDVPEGYKVSRKEKRNFPLILESPSPNQTSLYFCASSXXXXXYEQYFGPGTRLTVTEDLKNVFPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWWVNGKEVHSGVSTDPQPLKEQPALNDSRYCLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSESYQQGVLSATILYEILLGKATLYAVLVSALVLMAMVKRKDSRG\n",
      "5.0 10.0 CASSYQTGAIYGYTF ['TRBV6-5*01'] ['TRBJ1-2*01']\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHEYMSWYRQDPGMGLRLIHYSVGAGITDQGEVPNGYNVSRSTTEDFPLRLLSAAPSQTSVYF\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHEYMSWYRQDPGMGLRLIHYSVGAGITDQGEVPNGYNVSRSTTEDFPLRLLSAAPSQTSVYFCASSYXXXXXYGYTF\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHEYMSWYRQDPGMGLRLIHYSVGAGITDQGEVPNGYNVSRSTTEDFPLRLLSAAPSQTSVYFCASSYXXXXXYGYTFGSGTRLTVV\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHEYMSWYRQDPGMGLRLIHYSVGAGITDQGEVPNGYNVSRSTTEDFPLRLLSAAPSQTSVYFCASSYXXXXXYGYTFGSGTRLTVVEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFFPDHVELSWWVNGKEVHSGVSTDPQPLKEQPALNDSRYCLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSVSYQQGVLSATILYEILLGKATLYAVLVSALVLMAMVKRKDF\n",
      "5.0 9.0 CASSLAPGATNEKLFF ['TRBV7-9*01', 'TRBV6-1*01', 'TRBV7-8*01', 'TRBV7-6*01', 'TRBV11-1*01'] ['TRBJ1-4*01']\n",
      "DTGVSQNPRHKITKRGQNVTFRCDPISEHNRLYWYRQTLGQGPEFLTYFQNEAQLEKSRLLSDRFSAERPKGSFSTLEIQRTEQGDSAMYL\n",
      "DTGVSQNPRHKITKRGQNVTFRCDPISEHNRLYWYRQTLGQGPEFLTYFQNEAQLEKSRLLSDRFSAERPKGSFSTLEIQRTEQGDSAMYLCASSLXXXXTNEKLFF\n",
      "DTGVSQNPRHKITKRGQNVTFRCDPISEHNRLYWYRQTLGQGPEFLTYFQNEAQLEKSRLLSDRFSAERPKGSFSTLEIQRTEQGDSAMYLCASSLXXXXTNEKLFFGSGTQLSVL\n",
      "DTGVSQNPRHKITKRGQNVTFRCDPISEHNRLYWYRQTLGQGPEFLTYFQNEAQLEKSRLLSDRFSAERPKGSFSTLEIQRTEQGDSAMYLCASSLXXXXTNEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFFPDHVELSWWVNGKEVHSGVSTDPQPLKEQPALNDSRYCLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSVSYQQGVLSATILYEILLGKATLYAVLVSALVLMAMVKRKDF\n",
      "5.0 9.0 CASSLAPGATNEKLFF ['TRBV7-9*01', 'TRBV6-1*01', 'TRBV7-8*01', 'TRBV7-6*01', 'TRBV11-1*01'] ['TRBJ1-4*01']\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYF\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSEXXXXTNEKLFF\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSEXXXXTNEKLFFGSGTQLSVL\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSEXXXXTNEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFFPDHVELSWWVNGKEVHSGVSTDPQPLKEQPALNDSRYCLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSVSYQQGVLSATILYEILLGKATLYAVLVSALVLMAMVKRKDF\n",
      "5.0 9.0 CASSLAPGATNEKLFF ['TRBV7-9*01', 'TRBV6-1*01', 'TRBV7-8*01', 'TRBV7-6*01', 'TRBV11-1*01'] ['TRBJ1-4*01']\n",
      "GAGVSQSPRYKVAKRGQDVALRCDPISGHVSLFWYQQALGQGPEFLTYFQNEAQLDKSGLPSDRFFAERPEGSVSTLKIQRTQQEDSAVYL\n",
      "GAGVSQSPRYKVAKRGQDVALRCDPISGHVSLFWYQQALGQGPEFLTYFQNEAQLDKSGLPSDRFFAERPEGSVSTLKIQRTQQEDSAVYLCASSLXXXXTNEKLFF\n",
      "GAGVSQSPRYKVAKRGQDVALRCDPISGHVSLFWYQQALGQGPEFLTYFQNEAQLDKSGLPSDRFFAERPEGSVSTLKIQRTQQEDSAVYLCASSLXXXXTNEKLFFGSGTQLSVL\n",
      "GAGVSQSPRYKVAKRGQDVALRCDPISGHVSLFWYQQALGQGPEFLTYFQNEAQLDKSGLPSDRFFAERPEGSVSTLKIQRTQQEDSAVYLCASSLXXXXTNEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFFPDHVELSWWVNGKEVHSGVSTDPQPLKEQPALNDSRYCLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSVSYQQGVLSATILYEILLGKATLYAVLVSALVLMAMVKRKDF\n",
      "5.0 9.0 CASSLAPGATNEKLFF ['TRBV7-9*01', 'TRBV6-1*01', 'TRBV7-8*01', 'TRBV7-6*01', 'TRBV11-1*01'] ['TRBJ1-4*01']\n",
      "GAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQGPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYR\n",
      "GAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQGPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLXXXXTNEKLFF\n",
      "GAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQGPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLXXXXTNEKLFFGSGTQLSVL\n",
      "GAGVSQSPRYKVTKRGQDVALRCDPISGHVSLYWYRQALGQGPEFLTYFNYEAQQDKSGLPNDRFSAERPEGSISTLTIQRTEQRDSAMYRCASSLXXXXTNEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFFPDHVELSWWVNGKEVHSGVSTDPQPLKEQPALNDSRYCLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSVSYQQGVLSATILYEILLGKATLYAVLVSALVLMAMVKRKDF\n",
      "5.0 9.0 CASSLAPGATNEKLFF ['TRBV7-9*01', 'TRBV6-1*01', 'TRBV7-8*01', 'TRBV7-6*01', 'TRBV11-1*01'] ['TRBJ1-4*01']\n",
      "EAEVAQSPRYKITEKSQAVAFWCDPISGHATLYWYRQILGQGPELLVQFQDESVVDDSQLPKDRFSAERLKGVDSTLKIQPAELGDSAMYL\n",
      "EAEVAQSPRYKITEKSQAVAFWCDPISGHATLYWYRQILGQGPELLVQFQDESVVDDSQLPKDRFSAERLKGVDSTLKIQPAELGDSAMYLCASSLXXXXTNEKLFF\n",
      "EAEVAQSPRYKITEKSQAVAFWCDPISGHATLYWYRQILGQGPELLVQFQDESVVDDSQLPKDRFSAERLKGVDSTLKIQPAELGDSAMYLCASSLXXXXTNEKLFFGSGTQLSVL\n",
      "EAEVAQSPRYKITEKSQAVAFWCDPISGHATLYWYRQILGQGPELLVQFQDESVVDDSQLPKDRFSAERLKGVDSTLKIQPAELGDSAMYLCASSLXXXXTNEKLFFGSGTQLSVLEDLNKVFPPEVAVFEPSEAEISHTQKATLVCLATGFFPDHVELSWWVNGKEVHSGVSTDPQPLKEQPALNDSRYCLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRADCGFTSVSYQQGVLSATILYEILLGKATLYAVLVSALVLMAMVKRKDF\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/alex/anaconda2/envs/py36/lib/python3.6/site-packages/Bio/Seq.py:2309: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "with open(\"nlv-llw.sequences_out.txt\", \"w\") as out:\n",
    "    #out.write(\"cid\\tV.segm.seq\\tCDR3\\tJ.segm.seq\\tC.segm.seq\\ttype\\tV.segm\\tJ.segm\\tspecie\\n\")\n",
    "    for index, tcr in minmeanvertices.iterrows():\n",
    "        #In the future I'll implement tcr_construction as function on the columns\n",
    "        v_genes = tcr[\"from.v.segm\"].split(',')\n",
    "        j_genes = tcr[\"from.j.segm\"].split(',')\n",
    "        cdr3 = tcr[\"from.cdr3aa\"]\n",
    "        specie = tcr[\"species\"]\n",
    "        for v_gene in v_genes:\n",
    "            for j_gene in j_genes:\n",
    "                print(tcr[\"from.v.end\"], tcr[\"from.j.start\"], cdr3, v_genes, j_genes)\n",
    "                \n",
    "                v_end = tcr[\"from.v.end\"]\n",
    "                if v_end < 0:\n",
    "                    v_end = 1\n",
    "                j_start = tcr[\"from.j.start\"]\n",
    "                if j_start <0:\n",
    "                    print(j_start)\n",
    "                if v_end > j_start:\n",
    "                    j_start = v_end\n",
    "                \n",
    "                tcr_nucleseq = tcr_n_constructor.get_tcr(cdr3, specie, int(v_end), int(j_start), v_gene, j_gene, add_c=True, seqlengths=True)\n",
    "                tcr_aminoseq = tcr_a_constructor.get_tcr(cdr3, specie, v_gene, j_gene, add_c=True, seqlengths=True)\n",
    "                \n",
    "                nucleseq = tcr_nucleseq[0]\n",
    "                tcr_nucle_lengths = list(tcr_nucleseq[1:])\n",
    "                #row = [tcr[\"cid\"]]\n",
    "                v_end = tcr[\"from.v.end\"]\n",
    "                if v_end < 0:\n",
    "                    v_end = 1\n",
    "                j_start = tcr[\"from.j.start\"]\n",
    "                if j_start <0:\n",
    "                    print(j_start)\n",
    "                out.write(\">{}|{}|{}|{}|{}|{}|\\n\".format(tcr[\"cid\"],\n",
    "                                                      v_gene, \n",
    "                                                      j_gene, \n",
    "                                                      specie, \n",
    "                                                      \"nucleotide\", \n",
    "                                                      cdr3))\n",
    "                row=[]\n",
    "                for i in range(1, len(tcr_nucle_lengths)):\n",
    "                    tcr_nucle_lengths[i] += tcr_nucle_lengths[i-1]\n",
    "                tcr_nucle_lengths[0] -= int(v_end)*3\n",
    "                tcr_nucle_lengths[1] += (len(cdr3)-int(j_start))*3\n",
    "                for i in range(len(tcr_nucle_lengths)):\n",
    "                    if i == 0:\n",
    "                        print(str(Seq(nucleseq[:tcr_nucle_lengths[i]]).translate()))\n",
    "                        row.append(nucleseq[:tcr_nucle_lengths[i]].replace(\"N\", \"?\"))\n",
    "                    else:\n",
    "                        print(str(Seq(nucleseq[:tcr_nucle_lengths[i]]).translate()))\n",
    "                        row.append(nucleseq[tcr_nucle_lengths[i-1]:tcr_nucle_lengths[i]].replace(\"N\", \"?\"))\n",
    "                #row += [\"nucleotide\", v_gene, j_gene, specie]\n",
    "                out.write(\"{}\\n\\n\".format(\"\\t\".join(row)))\n",
    "                \n",
    "                \n",
    "                aminoseq = tcr_aminoseq[0]\n",
    "                tcr_amino_lengths = list(tcr_aminoseq[1:])\n",
    "                #row = [tcr[\"cid\"]]\n",
    "                row = []\n",
    "                out.write(\">{}|{}|{}|{}|{}|{}|\\n\".format(tcr[\"cid\"],\n",
    "                                                      v_gene, \n",
    "                                                      j_gene, \n",
    "                                                      specie, \n",
    "                                                      \"aminoacid\", \n",
    "                                                      cdr3))\n",
    "                for i in range(len(tcr_amino_lengths)):\n",
    "                    if i == 0:\n",
    "                        row.append(aminoseq[:tcr_amino_lengths[i]])\n",
    "                    else:\n",
    "                        tcr_amino_lengths[i] += tcr_amino_lengths[i-1]\n",
    "                        row.append(aminoseq[tcr_amino_lengths[i-1]:tcr_amino_lengths[i]])\n",
    "                #row += [\"aminoacid\", v_gene, j_gene, specie]\n",
    "                out.write(\"{}\\n\\n\\n\".format(\"\\t\".join(row)))\n",
    "\n",
    "                #print(\">{}\\n{}\\n{}\\n{}\\n\".format(tcr[\"cid\"], tcr_nucleseq[0], tcr_aminoseq[0], str(Seq(tcr_nucleseq[0], generic_dna).translate())))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check for mistakes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.0 10.0 CASSYQTGASYGYTF TRBV6-1*01 TRBJ1-2*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSY\n",
      "False\n",
      "\n",
      "\n",
      "5.0 10.0 CASSYQTGASYGYTF TRBV6-1*01 TRBJ1-2*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSY\n",
      "False\n",
      "\n",
      "\n",
      "5.0 10.0 CASSYQTGASYGYTF TRBV6-1*01 TRBJ1-2*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSY\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n",
      "5.0 9.0 CASSLAPGATNEKLFF TRBV6-1*01 TRBJ1-4*01\n",
      "Error in V part\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSE\n",
      "NAGVTQTPKFQVLKTGQSMTLQCAQDMNHNSMYWYRQDPGMGLRLIYYSASEGTTDKGEVPNGYNVSRLNKREFSLRLESAAPSQTSVYFCASSL\n",
      "False\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/alex/anaconda2/envs/py36/lib/python3.6/site-packages/Bio/Seq.py:2309: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "#out.write(\"cid\\tV.segm.seq\\tCDR3\\tJ.segm.seq\\tC.segm.seq\\ttype\\tV.segm\\tJ.segm\\tspecie\\n\")\n",
    "for index, tcr in sorted_graph.iterrows():\n",
    "    #In the future I'll implement tcr_construction as function on the columns\n",
    "    v_genes = tcr[\"from.v.segm\"].split(',')\n",
    "    j_genes = tcr[\"from.j.segm\"].split(',')\n",
    "    cdr3 = tcr[\"from.cdr3aa\"]\n",
    "    specie = tcr[\"species\"]\n",
    "    for v_gene in v_genes:\n",
    "        for j_gene in j_genes:\n",
    "            v_end = tcr[\"from.v.end\"]\n",
    "            if v_end < 0:\n",
    "                v_end = 1\n",
    "            j_start = tcr[\"from.j.start\"]\n",
    "            if j_start <0:\n",
    "                print(j_start)\n",
    "            if v_end > j_start:\n",
    "                j_start = v_end\n",
    "            \n",
    "            tcr_nucleseq = tcr_n_constructor.get_tcr(cdr3, specie, int(v_end), int(j_start), v_gene, j_gene, add_c=True, seqlengths=True)\n",
    "            tcr_aminoseq = tcr_a_constructor.get_tcr(cdr3, specie, v_gene, j_gene, add_c=True, seqlengths=True)\n",
    "\n",
    "            nucleseq = tcr_nucleseq[0]\n",
    "            tcr_nucle_lengths = list(tcr_nucleseq[1:])\n",
    "            #row = [tcr[\"cid\"]]\n",
    "            \n",
    "            aminoseq = tcr_aminoseq[0]\n",
    "            tcr_amino_lengths = list(tcr_aminoseq[1:])\n",
    "            \n",
    "            \n",
    "            for i in range(1, len(tcr_nucle_lengths)):\n",
    "                tcr_nucle_lengths[i] += tcr_nucle_lengths[i-1]\n",
    "            lenseq = 0\n",
    "            for i in range(len(tcr_nucle_lengths)):\n",
    "                if i == 0:\n",
    "                    seqtrans = str(Seq(nucleseq[:tcr_nucle_lengths[i]]).translate())\n",
    "                elif i == 3:\n",
    "                    seqtrans = str(Seq(nucleseq[tcr_nucle_lengths[i-1]-1:tcr_nucle_lengths[i]]).translate()).replace(\"*\", \"\")\n",
    "                    #lenseq -= 1\n",
    "                else:\n",
    "                    seqtrans = str(Seq(nucleseq[tcr_nucle_lengths[i-1]:tcr_nucle_lengths[i]]).translate())\n",
    "                \n",
    "                seqamino = aminoseq[lenseq:lenseq+len(seqtrans)]\n",
    "                lenseq += len(seqtrans)\n",
    "                if (i != 1 and seqtrans != seqamino):\n",
    "                    print(tcr[\"from.v.end\"], tcr[\"from.j.start\"], cdr3, v_gene, j_gene)\n",
    "                    print(\"Error in {} part\".format({0:\"V\", 1:\"CDR3\", 2:\"J\", 3:\"C\"}[i]))\n",
    "                    print(\"{}\\n{}\\n{}\\n\\n\".format(seqtrans, seqamino, seqtrans==seqamino))\n",
    "                    \n",
    "            #row += [\"nucleotide\", v_gene, j_gene, specie]\n",
    "\n",
    "            #row += [\"aminoacid\", v_gene, j_gene, specie]\n",
    "\n",
    "            #print(\">{}\\n{}\\n{}\\n{}\\n\".format(tcr[\"cid\"], tc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (3.6 from conda2)",
   "language": "python",
   "name": "python3.6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
